File: Problem543/solution.py
"""
Let n be the number of nodes in the tree

Time Complexity: O(n)
Space Complexity: O(n)

Solution with a recursive dfs

"""

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def diameterOfBinaryTree(self, root: TreeNode) -> int:
        self.res = 0
        def dfs(node):
            if not node:
                return 0
            
            left = dfs(node.left) 
            right = dfs(node.right)
            self.res = max(self.res, left+right)
            return 1 + max(left, right)
        dfs(root)
        return self.res
    
sol:Solution = Solution()
p = TreeNode(1, TreeNode(2, TreeNode(4), TreeNode(5)), TreeNode(3))
print(sol.diameterOfBinaryTree(p))

File: Problem543/solution.cpp
/* 
Let n be the number of nodes in the tree

Time Complexity: O(n)
Space Complexity: O(n)

Solution with a recursive dfs.

*/

#include <iostream>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    int res = 0;

    int dfs(TreeNode* node){
        if(node == nullptr){
            return 0;
        }
        int left = dfs(node->left);
        int right = dfs(node->right);
        res = max(res, left+right);
        return 1 + max(left, right);
    }

    int diameterOfBinaryTree(TreeNode* root) {
        dfs(root);
        return res;
    }
};

int main(){

    Solution sol = Solution();
    TreeNode* p = new TreeNode(1, new TreeNode(2, new TreeNode(4), new TreeNode(5)), new TreeNode(3));

    cout << sol.diameterOfBinaryTree(p) << endl;

    delete p->right;
    delete p->left;
    delete p;

}